### ISSUE TITLE
Investigate exposing Blender support for audio recording

### ISSUE THREAD

**Issue by britalmeida** [britalmeida/blender_addon_push_to_talk#17]
In the interest of saving people from having to install ffmpeg separately, it would be good to use Blender's own ffmpeg, but that is currently not possible.

Needs doing: look up what functionality Blender offers via ffmpeg or audaspace and how that is (not?) exposed to add-ons.
Volunteers willing to investigate are welcome.

**Comment by globglob3D** [comment]
> it would be good to use Blender's own ffmpeg, but that is currently not possible.

Why is that the case?

I need to record audaspace playback within Blender for my own addon which can generate audio from samples but not save it to a file. I thought of using Blender's ffmpeg as well. I haven't tried it yet as I was looking for easier alternatives but I don't think there are any.

Any pointers would be appreciated.

**Comment by britalmeida** [comment]
Using Blender's own ffmpeg would mean that users don't have to install ffmpeg separately, which is a hassle and people have varying degrees of success.
There are quite a few support requests here due to people not understanding how to install it, or it ended up in an unusual place, permissions, etc.
A significant part of this add-on is about managing to find ffmpeg and parsing any possible text output it gives. It would be way nicer for the users and for the add-on if Blender would expose an API for some ffmpeg capabilities.

Other add-ons could make use of exposed ffmpeg too, for example, to be able to re-encode/export a film edit using the exact same ffmpeg that blender already uses. Otherwise some sort of renderfarm setup is needed which installs ffmpeg and runs it from the command line.

For Push To Talk, besides just ffmpeg encoding, Blender itself would need to provide an interface for finding and selecting a microphone, which is quite a bit of work and not as generally useful for other add-ons.

I'm not familiar with audaspace (this is a TODO!)
But I think it does support writing to a file?
https://docs.blender.org/api/current/aud.html#aud.Sound.write

If that doesn't help, then you're also looking to use an external ffmpeg in your add-on as Push To Talk does?

**Comment by globglob3D** [comment]
Ok I thought there might be a way to get to Blender's ffmpeg without relying on the API somehow but from what I can see searching for "ffmpeg" in Blender's install folder this is not possible. Only thing it returns is preset python files that control ffmpeg render parameters through bpy.context.scene.render.ffmpeg but it's limited to what Blender expose for rendering, so no recording option here.

It seems that an API change would be needed to expose more functionality, as you said. But realistically all we need is for ffmpeg python module to be bundled with Blender and be importable in python, no?

audaspace does support writing to a file but not recording a stream of audio. So if you play multiple samples at different volume, varying pitch, etc, you can't save that as one file.

> If that doesn't help, then you're also looking to use an external ffmpeg in your add-on as Push To Talk does?

Yes, unless I can find a simpler cross platform solution for recording a stream of audio (coming from the "computer" rather than a mic but I believe this difference does not matter much).

**Comment by britalmeida** [comment]
I see, so you can't get the mix that audaspace does in realtime.
That sounds more like an audaspace limitation? If I understand correctly, you're looking to replicate the mix and processing of different samples with your own add-on (which you've already done?) and now you need to dump that stream into a file?

**Comment by globglob3D** [comment]
Yes it is an audaspace limitation. I'm using audaspace to create the stream and playback and all I need is a way to record that. 
I've tried other modules, that need to be bundled, such as pyo for generating, playbacking, and recording audio but it had its own quirks and crashed Blender randomly for stupid reasons. All other solutions are not cross platforms and would require building wheels from scratch or bundling a lot of different modules for each OS and a lot of code to make them work. I was really hoping for an easier solution.

If I find something I'll let you know.

For context this is an example and my code so far using audaspace:

https://www.youtube.com/watch?v=Eqh4kBXx9zM

```
import bpy
import aud
import os
from . import utils

class DroneSound:

    def __init__(self):
        self.device = aud.Device()
        self.device.volume = 0

        addon_dir = os.path.dirname(os.path.abspath(__file__))
        sound_dir = os.path.join(addon_dir, "sound")

        sound = aud.Sound(os.path.join(sound_dir, f"propellers.wav"))
        self.handle = self.device.play(sound)
        self.handle.volume = 0
        self.handle.loop_count = -1

    def update_propellers_sound(self, throttle):
        """Change pitch based on throttle (0 to 1)."""

        self.device.volume = bpy.context.scene.other_parameters.volume

        self.handle.volume = utils.map_range_clamped(throttle, 0.05, 1, 0.01, 1)
        self.handle.pitch = utils.map_range_clamped(throttle, 0.05, 1, 0.5, 2.5)
    
    def stop_sounds(self):
        self.device.stopAll()
```

**Comment by britalmeida** [comment]
Sorry, I wouldn't know more as I don't know audaspace well and I'm also not a sound expert.
It looks like a different use case than what Push To Talk needs.
